### PR TITLE
add support for dotnet generations

### PR DESCRIPTION
**Pull request by anurse**
/cc @lodejard @Eilon @davidfowl

### CODE REVIEW

**Comment by anurse** [comment]
I haven't done a full `build verify` pass yet, but I wanted to let you three see it first :)

File: src/Microsoft.Dnx.Tooling/Restore/NuGetDependencyResolver.cs [hunk 1]
```diff
@@ -79,8 +79,7 @@ private IEnumerable<LibraryDependency> GetDependencies(PackageInfo packageInfo,
 
             // TODO: Remove this when we do #596
             // ASP.NET Core isn't compatible with generic PCL profiles
-            if (string.Equals(targetFramework.Identifier, VersionUtility.AspNetCoreFrameworkIdentifier, StringComparison.OrdinalIgnoreCase) ||
-                string.Equals(targetFramework.Identifier, VersionUtility.DnxCoreFrameworkIdentifier, StringComparison.OrdinalIgnoreCase))
+            if (string.Equals(targetFramework.Identifier, VersionUtility.DnxCoreFrameworkIdentifier, StringComparison.OrdinalIgnoreCase))
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Tooling/Restore/NuGetDependencyResolver.cs]: This should be IsPackageBased

File: test/Microsoft.Dnx.Runtime.Tests/VersionUtilityFacts.cs [hunk 2]
```diff
@@ -103,6 +87,7 @@ public void FrameworksAreCompatible(string project, string package, bool compati
         [InlineData("dotnet", ".NETPlatform", "5.0")]
         [InlineData("dotnet10", ".NETPlatform", "1.0")]
         [InlineData("dotnet50", ".NETPlatform", "5.0")]
+        [InlineData("dotnet5.1", ".NETPlatform", "5.1")]
         [InlineData("dotnet60", ".NETPlatform", "6.0")]
         public void CanParseShortFrameworkNames(string shortName, string longName, string version)
```
Comment (davidfowl) [inline test/Microsoft.Dnx.Runtime.Tests/VersionUtilityFacts.cs]: Need more cases that show false,

dotnet 5.1 -> net46, false Etc

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/DotNetGenerationMapping.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.Versioning;
+
+namespace NuGet
+{
+    internal static class DotNetGenerationMapping
+    {
+        private static readonly Version _version45 = new Version(4, 5);
+        private static readonly Dictionary<FrameworkName, Version> _mappings = new Dictionary<FrameworkName, Version>()
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/DotNetGenerationMapping.cs]: _generationMappings

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/DotNetGenerationMapping.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            // netN (where N > 4.5) => dotnetY
+            else if(input.Identifier.Equals(VersionUtility.NetFrameworkIdentifier) && input.Version >= _version45)
+            {
+                yield return GetGeneration(input);
+            }
+        }
+
+        public static FrameworkName GetGeneration(FrameworkName input)
+        {
+            var version = _mappings[input];
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/DotNetGenerationMapping.cs]: More defensive here. What happens if we don't map?

Comment (anurse) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/DotNetGenerationMapping.cs]: I originally had more defensive code but a) it makes the generator above complicated and b) we only ever call GetGeneration with frameworks we **know** should have mappings. I'll make it private (should always have been so) and give defensive coding one more shot though, I have an idea that might work

**Comment by anurse** [comment]
I'm impressed that it actually works ;). I didn't even finish my build verify run. I guess I'm just that good :trollface:

File: test/Microsoft.Dnx.Runtime.Tests/VersionUtilityFacts.cs (hunk 2 above)
Comment (anurse) [inline test/Microsoft.Dnx.Runtime.Tests/VersionUtilityFacts.cs]: This is actually the wrong test for that but I get your idea :P

**Comment by anurse** [comment]
Build is fixed and everything is squashed. Please review @lodejard @davidfowl

**Comment by davidfowl** [comment]
This is in right?

**Comment by anurse** [comment]
Yeah, it's in

**Comment by borgdylan** [comment]
Is there some article about .NET Generations?

**Comment by davidfowl** [comment]
@borgdylan There will be soon

**Pull request closed without merging**